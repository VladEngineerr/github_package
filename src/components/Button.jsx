import PropTypes from 'prop-types';
import classnames from 'classnames';

export default function Button({
  className,
  design = 'primary',
  disabled = false,
  onClick,
  testID = 'xButton',
  text,
  type = 'button',
}) {
  return (
    <button
      className={classnames(
        'button classes here',
        { disabled: 'disabled' },
        design,
        className
      )}
      onClick={onClick}
      type={type}
      data-testid={testID}
      disabled={disabled}
    >
      {text}
    </button>
  );
}

Button.propTypes = {
  className: PropTypes.string, // accept className for placement
  design: PropTypes.oneOf(['primary', 'secondary', 'link']), // built-in styles
  disabled: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  testID: PropTypes.string,
  text: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['button', 'submit']), // only allowable types, button to ignore form submission
};
