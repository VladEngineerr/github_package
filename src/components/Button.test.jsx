import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from './Button';

describe('Button', () => {
  describe('snapshots', () => {
    test('render with defaults', () => {
      const { baseElement } = render(<Button text="Test" onClick={vi.fn} />);

      expect(baseElement).toMatchSnapshot();
    });

    test('render disabled ', () => {
      const { baseElement } = render(
        <Button text="Test" onClick={vi.fn} disabled={true} />
      );

      expect(baseElement).toMatchSnapshot();
    });

    test('render with props', () => {
      const { baseElement } = render(
        <Button
          text="Test"
          type="submit"
          className="foobar"
          design="secondary"
          testID="xFooBar"
          onClick={vi.fn}
        />
      );

      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('integration', () => {
    let user;

    beforeEach(() => {
      user = userEvent.setup();
    });

    afterEach(() => {
      user = null;
    });

    test('onClick', async () => {
      const onClick = vi.fn();

      render(<Button text="Test" onClick={onClick} />);

      const button = screen.getByTestId('xButton');

      await user.click(button);
    });
  });
});

