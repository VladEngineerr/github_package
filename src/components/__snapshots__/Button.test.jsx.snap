// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > snapshots > render disabled  1`] = `
<body>
  <div>
    <button
      class="button classes here disabled primary"
      data-testid="xButton"
      disabled=""
      type="button"
    >
      Test
    </button>
  </div>
</body>
`;

exports[`Button > snapshots > render with defaults 1`] = `
<body>
  <div>
    <button
      class="button classes here disabled primary"
      data-testid="xButton"
      type="button"
    >
      Test
    </button>
  </div>
</body>
`;

exports[`Button > snapshots > render with props 1`] = `
<body>
  <div>
    <button
      class="button classes here disabled secondary foobar"
      data-testid="xFooBar"
      type="submit"
    >
      Test
    </button>
  </div>
</body>
`;
